flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/statistics/ContinuousHistogram.scala

```diff
@@ -0,0 +1,315 @@
+    val wantedSum = (q * total).round.toInt
+    var currSum = count(getValue(0))
+
+    if (wantedSum < currSum) {
+      require(lower > -MaxValue, "Set a lower bound before proceeding")
+      return Math.sqrt(2 * wantedSum * Math.pow(getValue(0) - lower, 2) / getCounter(0)) + lower
+    }
+
+    /** Walk the histogram to find sums at every bin value
+      * As soon as you hit the interval where you should be
+      * Walk along the trapezoidal line
+      * This leads to solving a quadratic equation.
+      */
+    for (i <- 1 to bins - 1) {
+      val tmpSum = count(getValue(i))
+      if (currSum <= wantedSum && wantedSum < tmpSum) {
+        val neededSum = wantedSum - currSum
+        val a: Double = getCounter(i) - getCounter(i - 1)
+        val b: Double = 2 * getCounter(i - 1)
+        val c: Double = -2 * neededSum
+        if (a == 0) {
+          return getValue(i - 1) + (getValue(i) - getValue(i - 1)) * (-c / b)
+        } else {
+          return getValue(i - 1) +
+            (getValue(i) - getValue(i - 1)) * (-b + Math.sqrt(b * b - 4 * a * c)) / (2 * a)
+        }
+      } else {
+        currSum = tmpSum
+      }
+    }
+    require(upper < MaxValue, "Set an upper bound before proceeding")
+    // this means wantedSum > sum(getValue(bins-1))
+    // this will likely fail to return a bounded value.
+    // Make sure you set some proper limits on min and max.
+    getValue(bins - 1) + Math.sqrt(
+      Math.pow(upper - getValue(bins - 1), 2) * 2 * (wantedSum - currSum) / getCounter(bins - 1))
+  }
+
+  /** Returns the probability (and by extension, the number of points) for the value b
+    * Since this is a continuous histogram, return the cumulative probability at b
+    *
+    * @return Cumulative number of points at b
+    */
+  def count(b: Double): Int = {
+    require(bins > 0, "Histogram is empty")
+    if (b < lower) {
+      return 0
+    }
+    if (b >= upper) {
+      return total
+    }
+    /** Suppose x is the index with value just less than or equal to b
+      * Then, sum everything up to x-1
+      * Add half the value at x
+      * Find area of the trapezoid for x and the value b
+      */
+
+    val index = find(b) - 1
+    var m_b, s: Double = 0
+    if (index == -1) {
+      m_b = getCounter(index + 1) * (b - lower) / (getValue(index + 1) - lower)
+      s = m_b * (b - lower) / (2 * (getValue(index + 1) - lower))
+      return s.round.toInt
+    } else if (index == bins - 1) {
+      m_b = getCounter(index) +
+        (-getCounter(index)) * (b - getValue(index)) / (upper - getValue(index))
+      s = (getCounter(index) + m_b) *
+        (b - getValue(index)) / (2 * (upper - getValue(index)))
+    } else {
+      m_b = getCounter(index) + (getCounter(index + 1) - getCounter(index)) *
+        (b - getValue(index)) / (getValue(index + 1) - getValue(index))
+      s = (getCounter(index) + m_b) *
+        (b - getValue(index)) / (2 * (getValue(index + 1) - getValue(index)))
```

Comment:
I meant mapping area as in, taking ratio of area to the number of elements. The assumption is that elements are uniformly distributed between two different bins. This is why the line connecting m_x and m_y is assumed to be linear.